Objective-C/CBLArray.mm

```diff
@@ -59,21 +66,33 @@ - (instancetype) initWithCopyOfMArray: (const MArray<id>&)mArray
     self = [super init];
     if (self) {
         _array.initAsCopyOf(mArray, isMutable);
+        _sharedLock = [self _sharedLock];
     }
     return self;
 }
 
 
+- (NSObject*) _sharedLock {
+    id db = nil;
+    if (_array.context() != MContext::gNullContext)
+        db = ((DocContext*)_array.context())->database();
+    return db != nil ? db : self;
+}
+
+
 - (id) copyWithZone:(NSZone *)zone {
     return self;
 }
 
 
 - (CBLMutableArray*) mutableCopyWithZone:(NSZone *)zone {
-    return [[CBLMutableArray alloc] initWithCopyOfMArray: _array isMutable: true];
+    CBL_LOCK(self.sharedLock) {
```

Comment:
Better to do `CBL_LOCK(_sharedLock)`, here and below. I know I'm being obsessive about performance, but these accessors are important and it's best to add a little overhead as possible.